ide/idetop.ml

```diff
@@ -212,25 +212,19 @@ let goals () =
     if Proof_diffs.show_diffs () then begin
       let oldp = Stm.get_prev_proof ~doc (Stm.get_current_state ~doc) in
       let diff_goal_map = Proof_diffs.make_goal_map oldp newp in
-      let map_goal_for_diff ng = (* todo: move to proof_diffs.ml *)
-        try Evar.Map.find ng diff_goal_map  with Not_found -> ng
-      in
 
       let process_goal_diffs nsigma ng =
         let open Evd in
-        let og = map_goal_for_diff ng in
         let og_s = match oldp with
         | Some oldp ->
           let (_,_,_,_,osigma) = Proof.proof oldp in
-          Some { it = og; sigma = osigma }
+          (try Some { it = Evar.Map.find ng diff_goal_map; sigma = osigma } with Not_found -> None)
```

Comment:
Hi @jfehrle, by curiosity, is this `try with Not_found` needed because you want to possibly parse strings of the form `?X = ?Y` with unbound `?X` and `?Y`, or is this `try` needed to fix #8922?

Reply:
Short answer: no, this is not related to parsing.

Proof_diffs.make_goal_map uses a heuristic to match new goals to old goals by looking at the old and new proof objects, which works fine for the cases I tried.  One of the problems uncovered by this bug report is a goal that could not be matched.  In the example below (after the `refine`), `?Goal1` doesn't appear in the `Show Proof`, so it couldn't be matched.  My initial fix was to add the `try with Not_found` here, which would treat any such goals as new, which would show them as new, fully highlighting them.

A day or two later I discovered that Proof.all_goals doesn't include shelved goals.  Including those goals fixed the matching problem of this bug in a cleaner way.  Although `?Goal1` seems a bit of a strange corner case to me (not appearing in the proof object at all), I'm not sure that change covers all the possibilities.

I think I will change the code so it raises Diff_failure instead as a way of surfacing other unhandled cases if they indeed exist.  (Note that users can easily work around such anomalies by turning off diffs for the proof steps getting anomalies.)

```
eza < Show Proof.
(let X := fun H : ?Goal0 => ?Goal2@{y:=H} in ?Goal)

eza < Show Existentials.
Existential 1 = ?Goal : [X : forall x : ?Goal0, ?Goal1 |- True]
Existential 2 = ?Goal0 : [ |- Type] (shelved)
Existential 3 = ?Goal1 : [x : ?Goal0 |- Type] (shelved)
Existential 4 = ?Goal2 : [y : ?Goal0 |- ?Goal1@{x:=y}]
```

Reply:
OK, thanks for the explanation.